### PR TITLE
Bug 548149 - Allow min/max date to show the allowed range

### PR DESCRIPTION
**Pull request by sibeon**
in the Calendar and to limit wrong selection

This new feature allowes to set min and max date to avoid selecting invalide date.
handled min max in:
- text entry - only valid dates can be entered in the text field. If date is invalid and no previous date is selected the Datepicker will be opend.
- date picker - invalid dates are set to enabled=false, today button makes date visible if today is not valid.

Signed-off-by: Stefan Nöbauer <stefan.noebauer@kgu-consulting.com>

### CODE REVIEW

File: widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTime.java [hunk 1]
```diff
@@ -1168,990 +1196,1023 @@ void handleKey(Event event) {
...
+		this.maxDate = normDate(this.builder.getMaxDate());
+
+		if (picker != null) {
+			disposePicker();
+			createPicker();
+		}
+	}
+
+	private Calendar normDate(Calendar date) {
+		if (date != null) {
```
Comment (wimjongman) [inline widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTime.java]: Code smell ;) You alter the input parameter.

File: widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTime.java [hunk 2]
```diff
@@ -1168,990 +1196,1023 @@ void handleKey(Event event) {
...
+		}
+	}
+	
+	/** 
+	 * Returns the minimum date or <code>null</code>.
+	 * 
+	 * @return Returns the minDate.
+	 * @since 1.4.0
+	 */
+	public Calendar getMinDate() {
```
Comment (wimjongman) [inline widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTime.java:2196]: Allows mindate to be altered. Is this your intention? If so then state that in the javadoc. If not then return a clone of mindate.

File: widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTime.java [hunk 3]
```diff
@@ -1168,990 +1196,1023 @@ void handleKey(Event event) {
...
+		return minDate;
+	}
+
+	/** 
+	 * Returns the maximum date or <code>null</code>.
+	 * 
+	 * @return Returns the maxDate.
+	 * @since 1.4.0
+	 */
+	public Calendar getMaxDate() {
```
Comment (wimjongman) [inline widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTime.java:2209]: Allows maxdate to be altered. Is this your intention? If so then state that in the javadoc. If not then return a clone of maxdate.

File: widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTimeBuilder.java [hunk 4]
```diff
@@ -131,6 +136,20 @@ public boolean hasHeader(int type) {
 		return false;
 	}
 
+	/**
+	 * @return Returns the minDate.
+	 */
+	public Calendar getMinDate() {
```
Comment (wimjongman) [inline widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTimeBuilder.java:142]: Allows date to be altered. Is this your intention? If so then state that in the javadoc. If not then return a clone of date.

File: widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTimeBuilder.java [hunk 5]
```diff
@@ -131,6 +136,20 @@ public boolean hasHeader(int type) {
...
+		return minDate;
+	}
+	
+	/**
+	 * @return Returns the maxDate.
```
Comment (wimjongman) [inline widgets/cdatetime/org.eclipse.nebula.widgets.cdatetime/src/org/eclipse/nebula/widgets/cdatetime/CDateTimeBuilder.java]: Allows date to be altered. Is this your intention? If so then state that in the javadoc. If not then return a clone of date.

**Review by wimjongman — CHANGES_REQUESTED** [review]
Hi Stefan, Thanks for the patch. Please take a look at my comments.

**Comment by wimjongman** [comment]
jenkins, add to whitelist

**Comment by sibeon** [comment]
OK I pushed some changes.

**Comment by sibeon** [comment]
Hi
Can I do anything here to move forward?

**Comment by wimjongman** [comment]
Stefan, there is a build error. If you take a look at the log you can find what is wrong.

I think the CDateTime parent pom needs updating as well. You can find it one directory higher from the project.

**Comment by wimjongman** [comment]
retest this please

**Comment by sibeon** [comment]
> 
> 
> retest this please

Rebased on upstream/master and retested.

**Comment by wimjongman** [comment]
Thanks Stefan!

**Pull request merged**